src/function/scalar/map/map.cpp

```diff
@@ -66,46 +66,91 @@ void MapConversionVerify(Vector &vector, idx_t count) {
 	auto &value_vector = MapVector::GetValues(result);
-	auto list_data = ListVector::GetData(result);
+	auto result_data = ListVector::GetData(result);
 
+	result.SetVectorType(VectorType::CONSTANT_VECTOR);
 	if (args.data.empty()) {
 		ListVector::SetListSize(result, 0);
-		list_data->offset = 0;
-		list_data->length = 0;
+		result_data->offset = 0;
+		result_data->length = 0;
 		result.Verify(args.size());
 		return;
 	}
 
-	auto args_data = ListVector::GetData(args.data[0]);
+	bool keys_are_const = args.data[0].GetVectorType() == VectorType::CONSTANT_VECTOR;
+	bool values_are_const = args.data[1].GetVectorType() == VectorType::CONSTANT_VECTOR;
+	if (!keys_are_const || !values_are_const) {
+		result.SetVectorType(VectorType::FLAT_VECTOR);
+	}
+
 	auto key_count = ListVector::GetListSize(args.data[0]);
 	auto value_count = ListVector::GetListSize(args.data[1]);
-	if (key_count != value_count) {
-		throw InvalidInputException(
-		    "Error in MAP creation: key list has a different size from value list (%lld keys, %lld values)", key_count,
-		    value_count);
+	auto key_data = ListVector::GetData(args.data[0]);
+	auto value_data = ListVector::GetData(args.data[1]);
+	auto src_data = key_data;
+
+	if (keys_are_const && !values_are_const) {
+		Vector expanded_const(ListType::GetChildType(args.data[0].GetType()), value_count);
+
+		auto expansion_factor = value_count / key_count;
+		if (expansion_factor != args.size()) {
+			throw InvalidInputException("Error in MAP creation: key list and value list do not align. i.e. different "
+			                            "size or incompatible structure");
+		}
+		ExpandVector(args.data[0], expanded_const, expansion_factor);
+		key_vector.Reference(expanded_const);
+
+		src_data = value_data;
+	} else if (values_are_const && !keys_are_const) {
+		Vector expanded_const(ListType::GetChildType(args.data[1].GetType()), key_count);
+
+		auto expansion_factor = key_count / value_count;
+		if (expansion_factor != args.size()) {
+			throw InvalidInputException("Error in MAP creation: key list and value list do not align. i.e. different "
+			                            "size or incompatible structure");
+		}
+		ExpandVector(args.data[1], expanded_const, expansion_factor);
+		value_vector.Reference(expanded_const);
+	} else {
+		if (key_count != value_count || memcmp(key_data, value_data, args.size() * sizeof(list_entry_t)) != 0) {
+			throw InvalidInputException("Error in MAP creation: key list and value list do not align. i.e. different "
+			                            "size or incompatible structure");
+		}
 	}
-	ListVector::Reserve(result, key_count);
-	ListVector::SetListSize(result, key_count);
 
+	ListVector::SetListSize(result, MaxValue(key_count, value_count));
+
+	result_data = ListVector::GetData(result);
 	for (idx_t i = 0; i < args.size(); i++) {
-		list_data[i] = args_data[i];
+		result_data[i] = src_data[i];
+	}
+
+	if (!(keys_are_const && !values_are_const)) {
```

Comment:
Could you add additional comments explaining this section?
And possibly make the condition easier to read? I have to admit I don't know what this condition logically is